### PR TITLE
DATACOUCH-963 - Scope and collection api.

### LINKED ISSUE #963: Support for Scope and Collection API Changes [DATACOUCH-655]
**[Arun Vijayraghavan](https://jira.spring.io/secure/ViewProfile.jspa?name=JIRAUSER49896)** opened **[DATACOUCH-655](https://jira.spring.io/browse/DATACOUCH-655?redirect=false)** and commented

Not sure if this ticket is needed, but this will stand as a place holder for the API changes w.r.t underlying server specific scopes and collections API Changes.

 

Please feel free to close this ticket if you think this will be taken care by the JAVA SDK

---
No further details from [DATACOUCH-655](https://jira.spring.io/browse/DATACOUCH-655?redirect=false)

### PR DESCRIPTION
**Pull request by mikereiche**
Draft of changes for scope and collection API to share with Michael Nitschinger.
Some of the mechanisms in this change-set have been ruled out (the proliferation
of all the method signatures, for instance).

Closes #963
Original pull request: #1071.

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATACOUCH).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: src/main/java/org/springframework/data/couchbase/core/support/MapList.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package org.springframework.data.couchbase.core.support;
+
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+
+public class MapList<K,V> extends HashMap<K, List<V>> {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/couchbase/core/support/MapList.java:7]: There's no need for this one as Spring Framework ships with `MultiValueMap`.

File: src/main/java/org/springframework/data/couchbase/core/support/ScopeName.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package org.springframework.data.couchbase.core.support;
+
+public class ScopeName {
```
Comment (mp911de) [inline src/main/java/org/springframework/data/couchbase/core/support/ScopeName.java:3]: Where possible, use immutable value objects.

File: src/main/java/org/springframework/data/couchbase/core/support/CollectionName.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package org.springframework.data.couchbase.core.support;
+
+public class CollectionName {
+  String name;
+  public CollectionName(String name){
```
Comment (mp911de) [inline src/main/java/org/springframework/data/couchbase/core/support/CollectionName.java:5]: Where possible, use immutable value objects.

File: src/main/java/org/springframework/data/couchbase/repository/CouchbaseRepository.java [hunk 4]
```diff
@@ -32,15 +48,135 @@
 @NoRepositoryBean
 public interface CouchbaseRepository<T, ID> extends PagingAndSortingRepository<T, ID> {
 
+	// all the inherited methods need to be supplemented by ones that take Options, CollectionName and ScopeName args
+
+	@Override
+	Optional<T> findById(ID id);
+	Optional<T> findById(ID id, QueryOptions... options);
```
Comment (mp911de) [inline src/main/java/org/springframework/data/couchbase/repository/CouchbaseRepository.java:55]: That is exactly what we want to avoid. The repository must not accept store-specific types as we do not want the calling code to interact with store specifics.

Store specifics at the Template API level are fine.

File: src/main/java/org/springframework/data/couchbase/repository/CouchbaseRepository.java [hunk 5]
```diff
@@ -32,15 +48,135 @@
...
+	Optional<T> findById(ID id, CollectionName collectionName, ScopeName... scopeNames);
+
+	@Override
+	List<T> findAll();
+	List<T> findAll(QueryOptions... options);
+	List<T> findAll(QueryOptions options, CollectionName collectionName, ScopeName... scopeNames);
+	List<T> findAll(CollectionName collectionName, ScopeName... scopeNames);
+
+
+	List<T> findAll(org.springframework.data.couchbase.core.query.Query query);
```
Comment (mp911de) [inline src/main/java/org/springframework/data/couchbase/repository/CouchbaseRepository.java:66]: Introducing `Query` makes the Repository somewhat pointless as this method duplicates what's possible on the Template API level.

**Review by mp911de — CHANGES_REQUESTED** [review]
All repository-related changes violate the core ideas of Spring Data repositories to keep these as pure as possible. Introducing another 50 method overloads clutters the API and pulls in a level of abstraction that duplicates the Template API.

Specifying store-specific options can be done on the Template API.

File: src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java [hunk 6]
```diff
@@ -59,7 +59,7 @@ ClusterType type() {
 
 	@Override
 	TestClusterConfig _start() throws Exception {
-		bucketname = UUID.randomUUID().toString();
+		bucketname = "my_bucket"; // UUID.randomUUID().toString();
```
Comment (daschl) [inline src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java:62]: why is this hardcoded now?

File: src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java [hunk 7]
```diff
@@ -69,8 +69,9 @@ TestClusterConfig _start() throws Exception {
 						.build())
 				.execute();
 
-		if (postResponse.code() != 202) {
-			throw new Exception("Could not create bucket: " + postResponse + ", Reason: " + postResponse.body().string());
+		String reason = postResponse.body().string();
+		if (postResponse.code() != 202 && !(reason.contains("Bucket with given name already exists"))) {
```
Comment (daschl) [inline src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java:73]: this shouldn't exist with the uuid change above, right?

File: src/test/java/org/springframework/data/couchbase/util/JavaIntegrationTests.java [hunk 8]
```diff
@@ -107,6 +107,7 @@ public static void beforeAll() {
 		ApplicationContext ac = new AnnotationConfigApplicationContext(Config.class);
 		couchbaseTemplate = (CouchbaseTemplate) ac.getBean(COUCHBASE_TEMPLATE);
 		reactiveCouchbaseTemplate = (ReactiveCouchbaseTemplate) ac.getBean(REACTIVE_COUCHBASE_TEMPLATE);
+		System.err.println("JavaIntegrationTests.beforeAll() couchbaseTemplate: " + couchbaseTemplate);
```
Comment (daschl) [inline src/test/java/org/springframework/data/couchbase/util/JavaIntegrationTests.java:110]: should these be logged instead? because then you can turn it on off with the logger config, vs these will show up all the time

File: src/test/java/org/springframework/data/couchbase/repository/query/N1qlQueryCreatorTests.java [hunk 9]
```diff
@@ -75,6 +77,19 @@ void createsQueryCorrectly() throws Exception {
 		assertEquals(query.export(), " WHERE " + where(i("firstname")).is("Oliver").export());
 	}
 
+	@Test
+	void createsQueryFieldAnnotationCorrectly() throws Exception {
```
Comment (daschl) [inline src/test/java/org/springframework/data/couchbase/repository/query/N1qlQueryCreatorTests.java:81]: how is that related to scopes & collections?

File: src/test/java/org/springframework/data/couchbase/domain/Airport.java [hunk 10]
```diff
@@ -53,4 +56,16 @@ public String getIcao() {
...
+	public Airport withId(String id) {
+		return new Airport(id, this.iata, this.icao);
+	}
+
+	public Airport withIcao(String icao) {
+		return new Airport(this.getId(), this.iata, icao);
+	}
+
+	public Airport clearVersion() {
+		version = Long.valueOf(0);
```
Comment (daschl) [inline src/test/java/org/springframework/data/couchbase/domain/Airport.java:68]: 0L might also work?

File: src/test/java/org/springframework/data/couchbase/core/CouchbaseTemplateKeyValueIntegrationTests.java [hunk 11]
```diff
@@ -236,8 +236,22 @@ void insertById() {
...
+			} catch (Exception ofe) {
+				System.out.println(""+i+" caught: "+ofe);
+				couchbaseTemplate.removeByQuery(User.class).withConsistency(QueryScanConsistency.REQUEST_PLUS).all();
+				if (i == 4) {
+					throw ofe;
+				}
+				sleepSecs(1);
+			}
+		}
 		assertEquals(user, inserted);
```
Comment (daschl) [inline src/test/java/org/springframework/data/couchbase/core/CouchbaseTemplateKeyValueIntegrationTests.java:255]: these would also have to move inside the loop to check every iteration?

File: src/main/java/org/springframework/data/couchbase/repository/query/ReactiveN1qlRepositoryQueryExecutor.java [hunk 12]
```diff
@@ -40,7 +40,7 @@ public ReactiveN1qlRepositoryQueryExecutor(final ReactiveCouchbaseOperations ope
 		this.queryMethod = queryMethod;
 		this.namedQueries = namedQueries;
 		this.evaluationContextProvider = evaluationContextProvider;
-		throw new RuntimeException("Deprecated");
+		//throw new RuntimeException("Deprecated");
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/repository/query/ReactiveN1qlRepositoryQueryExecutor.java:43]: how does this relate to the change?

File: src/main/java/org/springframework/data/couchbase/repository/query/ReactiveCouchbaseRepositoryQuery.java [hunk 13]
```diff
@@ -36,14 +36,13 @@
...
-		super(queryMethod, operations, new SpelExpressionParser(), evaluationContextProvider);
+			final ReactiveCouchbaseQueryMethod queryMethod, final NamedQueries namedQueries) {
+		super(queryMethod, operations, new SpelExpressionParser(), QueryMethodEvaluationContextProvider.DEFAULT);
 		this.operations = operations;
 		this.queryMethod = queryMethod;
 		this.namedQueries = namedQueries;
-		this.evaluationContextProvider = evaluationContextProvider;
-		throw new RuntimeException("Deprecated");
+		this.evaluationContextProvider = QueryMethodEvaluationContextProvider.DEFAULT;
+		//throw new RuntimeException("Deprecated");
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/repository/query/ReactiveCouchbaseRepositoryQuery.java:45]: since the  constructor is changed, is  this an internal class?

File: src/main/java/org/springframework/data/couchbase/repository/query/N1qlRepositoryQueryExecutor.java [hunk 14]
```diff
@@ -47,7 +47,7 @@ public N1qlRepositoryQueryExecutor(final CouchbaseOperations operations, final C
 		this.queryMethod = queryMethod;
 		this.namedQueries = namedQueries;
 		this.evaluationContextProvider = evaluationContextProvider;
-		throw new RuntimeException("Deprecated");
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/repository/query/N1qlRepositoryQueryExecutor.java:50]: these worry me a little :)

File: src/main/java/org/springframework/data/couchbase/core/ExecutableExistsByIdOperationSupport.java [hunk 15]
```diff
@@ -32,17 +32,25 @@
...
-		return new ExecutableExistsByIdSupport(template, null);
+		return new ExecutableExistsByIdSupport(template, null, null, null);
 	}
 
 	static class ExecutableExistsByIdSupport implements ExecutableExistsById {
 
 		private final CouchbaseTemplate template;
+		private final String scope;
+		private final String collection;
+		private final GetOptions options;
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/ExecutableExistsByIdOperationSupport.java:43]: ExistsOptions?

**Review by daschl — CHANGES_REQUESTED** [review]

File: src/main/java/org/springframework/data/couchbase/SimpleCouchbaseClientFactory.java [hunk 16]
```diff
@@ -74,7 +74,7 @@ private SimpleCouchbaseClientFactory(final Supplier<Cluster> cluster, final Stri
 
 	@Override
 	public CouchbaseClientFactory withScope(final String scopeName) {
-		return new SimpleCouchbaseClientFactory(cluster, bucket.name(), scopeName);
+		return new SimpleCouchbaseClientFactory(cluster, bucket.name(), scopeName != null ? scopeName : getScope().name());
```
[the remark this answers is not in the document]
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/SimpleCouchbaseClientFactory.java:77]: If scopeName is null, 
 the old behavior would return a clientFactory with scope = _default
 the new behavior returns a clientFactory with the same scope as the "this". (yes, it could actually return "this" instead of a new object - that is a better implementation).
 This is to allow just using withScope(scopeName) without checking if scopeName is null.

File: src/main/java/org/springframework/data/couchbase/core/CouchbaseOperations.java [hunk 17]
```diff
@@ -44,4 +45,9 @@
 	 */
 	CouchbaseClientFactory getCouchbaseClientFactory();
 
+	/**
+	 * @param pseudoArgs - the metaArgs to set on the ThreadLocal field of the CouchbaseOperations
+	 */
+	void setThrdLocalArgs(PseudoArgs<?> pseudoArgs);
```
[the remark this answers is not in the document]
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/CouchbaseOperations.java:51]: Definitely.

File: src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java (hunk 6 above)
Comment (mikereiche) [reply src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java:62]: With the creation of multiple scopes and collections and the indexing, the initialization is getting long.  This now reuses the bucket, scopes, collections etc. if they already exist.

File: src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java (hunk 7 above)
Comment (mikereiche) [reply src/test/java/org/springframework/data/couchbase/util/UnmanagedTestCluster.java:73]: When reusing "my_bucket", 202(created) and "already exists" is acceptable, don't throw an exception.

File: src/test/java/org/springframework/data/couchbase/util/JavaIntegrationTests.java (hunk 8 above)
Comment (mikereiche) [reply src/test/java/org/springframework/data/couchbase/util/JavaIntegrationTests.java:110]: Yes. I'll look for other System.out and change to log.debug

File: src/main/java/org/springframework/data/couchbase/core/support/WithScope.java [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package org.springframework.data.couchbase.core.support;
+
+/**
+ * A common interface for all of Insert, Replace, Upsert that take options
+ *
+ * @author Michael Reiche
+ * @param <T> - the entity class
+ */
+public interface WithScope<T> {
```
[the remark this answers is not in the document]
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/support/WithScope.java:24]: Yes - it should match - InScope.  Changing these an all the  FindByQueryWithScope etc. interfaces to FindByQueryInScope etc.

File: src/main/java/org/springframework/data/couchbase/core/ExecutableExistsByIdOperationSupport.java (hunk 15 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/ExecutableExistsByIdOperationSupport.java:43]: Yes.  Similar issue with RemoveByQuery - it had RemoveOptions instead of QueryOptions - and in both cases the options were not passed in execution.   Going through all the  Reactive*OperationSupport classes to check for usage of options.

Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/ExecutableExistsByIdOperationSupport.java:43]: I added "options" tests that pass an options with timeout of 10 nanoseconds, resulting in a timeout.

File: src/main/java/org/springframework/data/couchbase/core/ExecutableFindByIdOperation.java [hunk 19]
```diff
@@ -18,9 +18,19 @@
...
+import org.springframework.data.couchbase.core.support.WithGetOptions;
 import org.springframework.data.couchbase.core.support.WithProjectionId;
+import org.springframework.data.couchbase.core.support.InScope;
 
+import com.couchbase.client.java.kv.GetOptions;
+
+/**
+ * Get Operations
+ *
+ * @author Christoph Strobl
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/ExecutableFindByIdOperation.java:31]: same here, I'll stop pointing them out below

File: src/main/java/org/springframework/data/couchbase/core/query/Query.java [hunk 20]
```diff
@@ -322,19 +325,25 @@ public String toN1qlRemoveString(ReactiveCouchbaseTemplate template, String coll
...
+		}
 		if (getParameters() != null) {
 			if (getParameters() instanceof JsonArray) {
 				options.parameters((JsonArray) getParameters());
 			} else {
 				options.parameters((JsonObject) getParameters());
 			}
 		}
+		if (scanConsistency == null) {
+			if (getScanConsistency() != null) {
```
[the remark this answers is not in the document]
Comment (daschl) [reply src/main/java/org/springframework/data/couchbase/core/query/Query.java:340]: still outstanding?

File: src/main/java/org/springframework/data/couchbase/core/support/PseudoArgs.java [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.couchbase.core.support;
+
+import org.springframework.data.couchbase.core.ReactiveCouchbaseTemplate;
+import org.springframework.data.couchbase.core.query.Query;
+
+public class PseudoArgs<OPTS> {
+	OPTS options;
```
[the remark this answers is not in the document]
Comment (daschl) [reply src/main/java/org/springframework/data/couchbase/core/support/PseudoArgs.java:22]: still outstanding?

File: src/main/java/org/springframework/data/couchbase/core/support/PseudoArgs.java [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+	 * @@return the collection name
+	 */
+	public String getCollection() {
+		return this.collectionName;
+	}
+
+	/**
+	 * @@return the options from the ThreadLocal field of the template
+	 */
+	private OPTS getTlOptions(ReactiveCouchbaseTemplate template) {
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/support/PseudoArgs.java:100]: can those be also renamed to getThreadLocalOptions so it's not as confusing?

File: src/main/java/org/springframework/data/couchbase/repository/query/N1qlRepositoryQueryExecutor.java (hunk 14 above)
Comment (daschl) [reply src/main/java/org/springframework/data/couchbase/repository/query/N1qlRepositoryQueryExecutor.java:50]: ?

File: src/main/java/org/springframework/data/couchbase/repository/query/ReactiveCouchbaseRepositoryQuery.java (hunk 13 above)
Comment (daschl) [reply src/main/java/org/springframework/data/couchbase/repository/query/ReactiveCouchbaseRepositoryQuery.java:45]: ?

File: src/main/java/org/springframework/data/couchbase/repository/query/ReactiveN1qlRepositoryQueryExecutor.java (hunk 12 above)
Comment (daschl) [reply src/main/java/org/springframework/data/couchbase/repository/query/ReactiveN1qlRepositoryQueryExecutor.java:43]: ?

**Review by daschl — CHANGES_REQUESTED** [review]
found a couple things - also I noticed many of my questions did not have comments/answers, would be good  to get clarifications  on those as well (just look through the change they will pop up)
